src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2647 @@
+{
+    /* Too small. */
+    if (input_len < DNP3_CRC_LEN) {
+        return -1;
+    }
+
+    /* Get the number of complete blocks. */
+    int blocks = input_len / (DNP3_BLOCK_SIZE + DNP3_CRC_LEN);
+
+    /* And the number of bytes in the last block. */
+    int rem = input_len - (blocks * (DNP3_BLOCK_SIZE + DNP3_CRC_LEN));
+
+    if (rem) {
+        if (rem < DNP3_CRC_LEN) {
+            return -1;
+        }
+        return (blocks * DNP3_BLOCK_SIZE) + (rem - DNP3_CRC_LEN);
+    }
+    else {
+        return (blocks * DNP3_BLOCK_SIZE);
+    }
+}
+
+/**
+ * \brief Reassemble the application layer by stripping the CRCs.
+ *
+ * Remove the CRCs from the user data blocks.  The output is the user
+ * data with the CRCs removed as well as the transport header removed,
+ * but the input data still needs to include the transport header as
+ * its part of the first user data block.
+ *
+ * If the output length passed in is non-null, the new input data will
+ * be appended, and the output length pointer incremented as needed.
+ *
+ * \param input Input buffer starting at the transport header (which
+ *    will be removed from the output).
+ * \param input_len Length of the input buffer.
+ * \param output Pointer to output buffer (may be realloc'd).
+ * \param output_len Pointer to output length.
+ *
+ * \retval 1 if reassembly was successful, otherwise 0.
+ */
+static int DNP3ReassembleApplicationLayer(const uint8_t *input,
+    uint32_t input_len, uint8_t **output, uint32_t *output_len)
+{
+    int len = DNP3CalculateTransportLengthWithoutCRCs(input_len);
+
+    if (len <= 0) {
+        return 0;
+    }
+
+    /* Remove one byte for the transport header and make sure we have
+     * at least one byte of user data. */
+    if (--len < 1) {
+        return 0;
+    }
+
+    if (*output == NULL) {
+        *output = SCCalloc(1, len);
+        if (unlikely(*output == NULL)) {
+            return 0;
+        }
+    }
+    else {
+        uint8_t *ptr = SCRealloc(*output, (size_t)(*output_len + len));
+        if (unlikely(ptr == NULL)) {
+            return 0;
+        }
+        *output = ptr;
+    }
+
+    int offset = 0, block_size;
+    while ((uint32_t)offset < input_len) {
+        if (input_len - offset > DNP3_BLOCK_SIZE + DNP3_CRC_LEN) {
+            block_size = DNP3_BLOCK_SIZE + DNP3_CRC_LEN;
+        }
+        else {
+            block_size = input_len - offset;
+        }
+
+        /* If handling the first block (offset is 0), trim off the
+         * first byte which is the transport header, and not part of
+         * the application data. */
+        if (offset == 0) {
+            offset++;
+            block_size--;
+        }
+
+        /* Need at least 3 bytes to continue. One for application
+         * data, and 2 for the CRC.  If not, return failure for
+         * malformed frame. */
+        if (block_size < DNP3_CRC_LEN + 1) {
+            SCLogNotice("Bailing.");
```

Comment:
Converted to debug with a better message.  If this function returns 0 for failure, a malformed event is set.